packages/block-editor/src/hooks/use-bindings-attributes.js

```diff
@@ -22,89 +24,232 @@ import { unlock } from '../lock-unlock';
+ */
+export function canBindAttribute( blockName, attributeName ) {
+	return (
+		canBindBlock( blockName ) &&
+		BLOCK_BINDINGS_ALLOWED_BLOCKS[ blockName ].includes( attributeName )
+	);
+}
+
+/**
+ * This component is responsible for detecting and
+ * propagating data changes from the source to the block.
+ *
+ * @param {Object}   props                   - The component props.
+ * @param {string}   props.attrName          - The attribute name.
+ * @param {any}      props.attrValue         - The attribute value.
+ * @param {Object}   props.blockProps        - The block props with bound attribute.
+ * @param {Object}   props.source            - Source handler.
+ * @param {Object}   props.args              - The arguments to pass to the source.
+ * @param {Function} props.onPropValueChange - The function to call when the attribute changes.
+ * @return {null}                     This is a data-handling component. Render nothing.
+ */
+const BindingConnector = ( {
+	args,
+	attrName,
+	attrValue,
+	blockProps,
+	onPropValueChange,
+	source,
+} ) => {
+	const { placeholder, value: propValue } = source.useSource(
+		blockProps,
+		args
+	);
+
+	const blockName = blockProps.name;
+
+	const updateBoundAttibute = useCallback(
+		( newAttrValue, prevAttrValue ) => {
+			/*
+			 * If the attribute is a RichTextData instance,
+			 * (core/paragraph, core/heading, etc.)
+			 * convert it to HTML string.
+			 *
+			 * To do: it looks like a workaround.
+			 * Consider improving the attribute and metadata fields types.
+			 */
+			if ( prevAttrValue instanceof RichTextData ) {
+				prevAttrValue = prevAttrValue.toHTMLString();
 			}
 
-			return (
+			// Do not update the attribute if it's the same value.
+			if ( newAttrValue === prevAttrValue ) {
+				return;
+			}
+
+			onPropValueChange( { [ attrName ]: newAttrValue } );
+		},
+		[ attrName, onPropValueChange ]
+	);
+
+	useEffect( () => {
+		if ( typeof propValue !== 'undefined' ) {
+			updateBoundAttibute( propValue, attrValue );
+		} else if ( placeholder ) {
+			/*
+			 * Placeholder fallback.
+			 * If the attribute is `src` or `href`,
+			 * a placeholder can't be used because it is not a valid url.
+			 * Adding this workaround until
+			 * attributes and metadata fields types are improved and include `url`.
+			 */
+			const htmlAttribute =
+				getBlockType( blockName ).attributes[ attrName ].attribute;
+
+			if ( htmlAttribute === 'src' || htmlAttribute === 'href' ) {
+				updateBoundAttibute( null );
+				return;
+			}
+
+			updateBoundAttibute( placeholder );
+		}
+	}, [
+		updateBoundAttibute,
+		propValue,
+		attrValue,
+		placeholder,
+		blockName,
+		attrName,
+	] );
+
+	return null;
+};
+
+function BlockBindingBridge( { bindings, props, onPropValueChange } ) {
+	const { getBlockBindingsSource } = useSelect( ( select ) => {
+		return {
+			getBlockBindingsSource: unlock( select( blockEditorStore ) )
+				.getBlockBindingsSource,
+		};
+	}, [] );
```

Comment:
Not sure what is better but I just realized that we could get all the registered sources and avoid calling the `getBlockBindingsSource` function multiple times.

I did something similar in [this commit](https://github.com/WordPress/gutenberg/pull/59283/commits/0d0277e4c67e819316abe763e4a3c9b78fdf63e8).

Reply:
cc @michalczaplinski since we've moved the code here

Reply:
Yeah, what Mario suggested is slightly more performant! 👍

Reply:
Addressed here https://github.com/WordPress/gutenberg/pull/58895/commits/d91afe6f6ccd7bd4e1e44308289e7f19686ab27d